crypto/fipsmodule/hmac/hmac.c

```diff
@@ -120,119 +175,114 @@ void HMAC_CTX_free(HMAC_CTX *ctx) {
 
 int HMAC_Init_ex(HMAC_CTX *ctx, const void *key, size_t key_len,
                  const EVP_MD *md, ENGINE *impl) {
-  // We have to avoid the underlying SHA services updating the indicator
-  // state, so we lock the state here.
-  FIPS_service_indicator_lock_state();
-  int ret = 0;
-  if (md == NULL) {
-    md = ctx->md;
+  // TODO: Figure out how we want to handle engines in this case.
+  assert(impl == NULL);
+
+  if (md && ctx->md != md) {
+    ctx->methods = GetInPlaceMethods(md);
+    if (ctx->methods == NULL) {
+      // Unsupported md
+      return 0;
+    }
+    ctx->md = md;
+  } else if (ctx->initialized != 1) {
+    return 0;
   }
 
-  // If either |key| is non-NULL or |md| has changed, initialize with a new key
-  // rather than rewinding the previous one.
-  //
-  // TODO(davidben,eroman): Passing the previous |md| with a NULL |key| is
-  // ambiguous between using the empty key and reusing the previous key. There
-  // exist callers which intend the latter, but the former is an awkward edge
-  // case. Fix to API to avoid this.
-  if (md != ctx->md || key != NULL) {
-    uint8_t pad[EVP_MAX_MD_BLOCK_SIZE];
-    uint8_t key_block[EVP_MAX_MD_BLOCK_SIZE];
-    unsigned key_block_len;
-
-    size_t block_size = EVP_MD_block_size(md);
-    assert(block_size <= sizeof(key_block));
+  // At this point we know we have valid methods and a context allocated.
+  const HmacMethods *methods = ctx->methods;
+  size_t block_size = EVP_MD_block_size(methods->evp_md);
+  assert(block_size % 8 == 0);
+  assert(block_size <= EVP_MAX_MD_BLOCK_SIZE);
+
+  FIPS_service_indicator_lock_state();
```

Comment:
I suggest keeping the comment: " 
// We have to avoid the underlying SHA services updating the indicator
// state, so we lock the state here."

Reply:
I'll restore it.